buildDenoPackage: init

aMOPel opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

**This is a reopen of the reverted PR https://github.com/NixOS/nixpkgs/pull/407434**

EDIT:

This PR is ready for review. 

You don't need to read the PR the older comments below, as that was mainly about me understanding nixpkgs' requirements and emilazy explaining them to me

You can build the tests locally with 
```sh
nix build -f . --option allow-import-from-derivation false tests.build-deno-package
```
but mind that it will take a long time, since this PR adapts the `pkgs.deno` package, which means you have to fully rebuild it.

Compared to the previous PR, mainly the custom fetcher changed.

To understand what is going on in the fetcher, you can read this documentation I wrote about the features, the fetcher needs to have:

https://github.com/NixOS/nixpkgs/blob/bd54b5e47f0a7858336b38fa30aaea7541841eb5/pkgs/build-support/deno/fetch-deno-deps/readme.md

And to understand the general architecture of the fetcher, you can read this documentation, which is also a collection of general problems that language build helpers need to solve, along with a proposal for a standardized architecture, which could be improved upon and added to the nixpkgs documentation:

https://github.com/NixOS/nixpkgs/blob/bd54b5e47f0a7858336b38fa30aaea7541841eb5/pkgs/build-support/deno/fetch-deno-deps/generic-fetcher-architecture-docs.md

EDITEND


<details>
<summary>Origonal text</summary>

## Discussion

We need to discuss specifics.

@emanueljg
@06kellyjac
@emilazy
@hsjobeki 

### Previous Problem

@emilazy brought up concerns about the previous PR. 

Namely it's an issue that FODs are built using the Deno CLI as a dependency.
Since the Deno CLI considers the format of the dependency cache as an implementation detail, when we bump the Deno CLI version in nixpkgs, the Deno CLI could suddenly produce a FOD with a different hash. However nix caches FODs and until somebody would manually bump the output hash of the FOD, the cached version would be used, even if a rebuild would produce a different result. This can produce a mass breakage with a large delay, which can be very hard to debug.

The issue is fundamental to nix's FODs. The inputs (in this case the deno cli) are intentionally not checked when it comes to cache invalidation.

This is why build helpers in nixpkgs implement custom fetchers. This way nixpkgs has full control over when the FOD will change. If the expected dependency cache format from the Deno CLI would change, we would still produce the same FOD,
however the Deno CLI would tell us, that the format is invalid and throw an error. Then the custom fetcher will have to be adapted.

### Dependency cache format viewed as an interface

I see 3 parts to this Interface

1. The lock file has a version, currently `version: 5`. So the lock file is "versioned".

2. The format of how the directory tree for the dependencies looks like, that deno expects, is not stable or documented.

3. On top of that deno injects many stateful files into that directory tree, to cache work.

    3.1. Some files can simply be deleted and will be regenerated by the deno cli.
    3.2. Some files can't be deleted (`registry.json` and `meta.json` files), since Deno won't recognize the dependencies, if they are missing.

### Statement from Deno maintainers
I asked the maintainers in the Deno discord about the (in)stability of the interface.

> That's correct, it's a moving target and an implementation detail, we don't give any stability guarantees around it, because we need to be able to change that installation structure if needed.
>      --bartlomieju

> there's no documentation. It's somewhat stable but kind of unstable. I'd recommend looking into using the Rust crates like https://github.com/denoland/deno_cache_dir and https://crates.io/crates/deno_npm_cache
>      --dsherret

### Breaking changes survey

I surveyed previous versions of Deno and compared the interface pairwise:

**Versions:**

- deno	1.46.3	[ab7b6889ae9d484eed2876868209e33eb262511d](https://lazamar.co.uk/nix-versions/?package=deno&version=1.46.3&fullName=deno-1.46.3&keyName=deno&revision=ab7b6889ae9d484eed2876868209e33eb262511d&channel=nixpkgs-unstable#instructions)	2024-09-26

- deno	2.1.1	[882842d2a908700540d206baa79efb922ac1c33d](https://lazamar.co.uk/nix-versions/?package=deno&version=2.1.1&fullName=deno-2.1.1&keyName=deno&revision=882842d2a908700540d206baa79efb922ac1c33d&channel=nixpkgs-unstable#instructions)	2024-11-25

- deno	2.2.4	[c5dd43934613ae0f8ff37c59f61c507c2e8f980d](https://lazamar.co.uk/nix-versions/?package=deno&version=2.2.4&fullName=deno-2.2.4&keyName=deno&revision=c5dd43934613ae0f8ff37c59f61c507c2e8f980d&channel=nixpkgs-unstable#instructions)	2025-03-25

- deno	2.2.12	[4684fd6b0c01e4b7d99027a34c93c2e09ecafee2](https://lazamar.co.uk/nix-versions/?package=deno&version=2.2.12&fullName=deno-2.2.12&keyName=deno&revision=4684fd6b0c01e4b7d99027a34c93c2e09ecafee2&channel=nixpkgs-unstable#instructions)	2025-05-24

- deno	2.3.6 3078b9a9e75f1790e6d6ef9955fdc6a2d1740cc6 2025-06-23

**Diffs:**
- v1.46.3_v2.1.1: point 1. changed, point 2. changed
- v2.1.1_v2.2.4: no changes
- v2.2.4_v2.2.12: no changes
- v2.2.12_v2.3.6: point 1. changed, point 3.2 changed

The changes were all not monumental, so it could be feasible to keep a custom fetcher up to date.

Before v2.0.0, the deno cli worked quite differently. For example, `deno install` would not generate a lock file. This highlights an aspect that I didn't think about before. When using the Deno CLI in the FOD, the CLI itself is also something that could experience breaking changes.

### Building a custom fetcher

It would be possible to circumvent building the custom fetcher, if we make the deno version part of the `denoDeps` build's `name`. That way, when the deno cli version changes, the FOD has to be rebuilt and we don't run into the cache issue. However we will have to do more rebuilds of the `denoDeps` FOD's than necessary, since it appears that not every version bump does break the interface.

Implementing the fetcher could prove difficult due to the complexity of Deno's package system. Deno supports multiple package sources. 

1. `jsr:` from the jsr registry
2. `npm:` from the npm registry
3. A URL, directly fetching from a url, e.g. github or a cdn.

Each of these sources has their own distinct format.
NPM has a special role here, since NPM package generally can execute scripts after being installed to do some post-install work, like downloading some big asset. This also has to happen inside the FOD.
Unfortunately, we can't just reuse the `buildNpmPackage` build helper for that part without extra effort, since it expects a `package-lock.json` file, and we just have the `deno.lock` file, and conversion with transitive dependencies could prove difficult from what I have glimpsed so far.

What do you think about this?

Should I implement the custom fetcher or is it not worth it?

The way I see it, the maintenance burden is probably gonna be larger with the custom fetcher, since it will rely on more implementation details from Deno. And the upfront cost to implement the custom fetcher could be large.

</details>


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->


- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.


<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->
<!--
---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc
-->

aMOPel commented:
Deno maintainer dsherret pointed me to some rust crates, which the deno cli uses, like https://github.com/denoland/deno_cache_dir and https://crates.io/crates/deno_npm_cache
So I thought, it seems very reasonable to just write the fetcher in rust and use those crates as to get the logic straight from the source. Then I thought, since the deno cli is using those crates, I might as well look into the codebase of deno and copy the relevant code for the fetcher. At that point I am just pointlessly duplicating the deno cli, though.

Here comes the idea:
What if our custom fetcher is the deno cli, **BUT** not the version pinned in nixpkgs, but a different (older) one, just used for the fetcher.
This way we have decoupled the deno versions. And the problem you were worried about, @emilazy, could not occur anymore.
When `pkgs.deno` breaks compatibility with the current FODs dependency cache format, then we have to update the hashes.
We could enforce having to update the hashes, by making the deno fetcher version part of the FOD's name. Either way, there will be an error, since the old incompatible FODs were created with the old incompatible deno cli.

What do you think about that? Am I missing something?

aMOPel commented:
> we would have to keep the fixed Deno version and all its dependencies building forever, likely multiple versions as we add more to support newer lock file formats, and run into the fact that the versions will inevitably become EOL and develop security vulnerabilities (that may even be relevant to FOD fetching – e.g. TLS issues)

And that is not the case for a custom fetcher, because we can bump the dependencies of the custom fetcher (like the compiler) without changing the hashes of the FODs it produces?

---

Anyway, the perk of breaking hashes way less often, by decoupling the fetcher and transformer, convinced me of the solution.

Thanks you for your patience, I needed a little bit of time to fully understand the implications.

I'm gonna write the custom fetcher with go, since I'm not really acquainted with rust yet.

---

I also looked into a way to have an equivalent to `importNpmLock`, but so far it looks like it's not possible, due to how jsr and deno work together. Will keep you updated.

aMOPel commented:
So my plan currently is to just build an `importLock` functionality, no custom fetcher using a hash.

And it works like this:

_NOTE: since there are 3 kinds of dependencies with deno (npm, jsr and direct url), there are usually 3 cases to consider_

1. parse the lockfile into a common format (preprocessing step)
   1. get a `deno.lock` file as an argument.
   2. there is a parser (written in nix) for the lockfile that
      a. reads the version of the lockfile
      b. chooses the appropriate converter for the version
      c. converts the packages information from the lockfile into our custom `common lockfile format`
   3. there is a npm lockfile converter that converts the npm section of the `common lockfile format` into a format that npm accepts
2. fetch deps (FOD step)
   1. there is a custom fetcher for jsr dependencies. it receives the `common lockfile format` as an argument and
      a. fetches the files (using nixpkgs fetchurl), where each file needs its own derivation, since that is what we have the hashes for
      b. the derivations are added to the lockfile attrset
   2. there is a custom fetcher for url dependencies. it receives the `common lockfile format` as an argument and
      a. fetches the files (using nixpkgs fetchurl), where each file needs its own derivation, since that is what we have the hashes for
      b. the derivations are added to the lockfile attrset
   3. the generated npm lock file is fed to the `fetchNpmDeps` fetcher and the resulting node_modules folder is our `common npm deps folder structure`
3. convert folder structure and metadata (postprocessing step)
   1. there are 3 converters, they run inside a mkDerication buildPhase
      a. there is a `common jsr deps folder structure` converter
      b. there is a `common url deps folder structure` converter
      c. there is a `common npm deps folder structure` converter
   2. they convert our common formats into the formats that the current deno version expects, they also creates any necessary meta data files etc.
      this could be more involved logic, using information from the lockfile, which might be difficult to do well in just shell script, so this should probably be done in some stronger language.

With preprocessing and postprocessing, the FOD's won't change often, if ever, since they are decoupled from any breaking changes that deno could introduce.
If the lockfile changes, then we write a new preprocessing converter.
If the folder strucuture changes, then we write a new postprocessing converter.

Still it's possible that something in step 2 will have to be changed, so
the version of that fetcher should be part of the name of the FODs.

---

I am unsure about wether 2.3. is a good idea, since it introduces a dependency on the `fetchNpmDeps` helper.

You said

> but for NPM maybe we could adapt the existing fetcher codebase so that there’s a lower layer that can take the relevant information for a single package directly rather than requiring it to be in package-lock.json format

But I'm not sold on that yet either.

---

There are multiple individual rust crates that contain the logic for the things we need to do

For deno.lock: https://crates.io/crates/deno_lockfile
For npm cache: https://crates.io/crates/deno_npm_installer https://crates.io/crates/deno_npm_cache

and some more.

From what I can tell they are not really doing what we need.

On paper it seems like a good idea to use them, but somehow they don't really fit into the architecture I lined out now.

---

Ideas? Objections?

hsjobeki commented:
If we come up with a generic fetcher interface for npm and cdn/jsr dependencies there is still the gap of the custom files and metadata both package managers expect. Did you manage to figure out some way of doing that? Because from my understanding thats equally complex. 

I imagine deno could have a cli command to pre-populate its cache with offline packages, then run deno install or similar and it would produce its own structure as needed (In a non fod) as an idea.

Since you seemed to be quite active with the deno maintainers that could be worth asking

aMOPel commented on pkgs/by-name/de/deno/package.nix:
```diff
@@ -204,7 +204,7 @@ rustPlatform.buildRustPackage (finalAttrs: {
   postInstall = ''
     # Remove non-essential binaries like denort and test_server
-    find $out/bin/* -not -name "deno" -delete
+    find $out/bin/* -not -name "deno" -not -name "denort" -delete
```
> Deno already builds the denort binary but it's removed. The correct solution for us would probably have been to add an extra output & keep denort.

@06kellyjac is this what you meant

aMOPel replied:
> Deno already builds the denort binary but it's removed. The correct solution for us would probably have been to add an extra output & keep denort.

@06kellyjac is this what you meant

aMOPel commented:
@hsjobeki

> Thats what i ment with the original comment. How do you think of solving this? My proposal was to try using deno itself, to avoid complexity, unsure if deno has some interfaces to support it.

As I said in my comment before:

> Last week I finally managed to do exactly that for the jsr and http packages using deno's provided [library](https://jsr.io/@deno/cache-dir/doc/~/HttpCache.prototype.set). It took a while to find it and how to use it, though.

This interface only works for https and jsr imports, though. Yesterday @rorosen helped me to write a small rust wrapper around that interface, which you can look at here: https://github.com/NixOS/nixpkgs/pull/419255/files#diff-657a96d5ebec4ff9ca7cf304c40ec07d4eb6f05b12b910577aa4719377f14e10

I had a wrapper written in deno before, but it proved difficult packaging that wrapper cleanly with nix, without this very PR (chicken and egg problem).

 Currently, I handle npm imports manually. Which means there is also only limited support.
 
 I outlined all the features that I am aware of, both missing and implemented, in the readme, I wrote for the fetcher
 
 https://github.com/NixOS/nixpkgs/pull/419255/files#diff-058bd94aa5598ce3d721e602626409d4e0713ae6ac10433d393d81e93263a67f

aMOPel commented:
@emilazy thanks for checking in

> We unfortunately can’t import locks from fetched sources because that would be IFD (import‐from‐derivation – a misleading term for “evaluation depending on the contents of derivation outputs”), and in general people aren’t too happy about having to vendor lock files in Nixpkgs to work around this (https://github.com/NixOS/nixpkgs/issues/327064). So I’m not sure what the current state of things is, but I would suggest prioritizing a hash‐based fetcher, as it’s likely what we’d want to use for most packages that have lock files checked in upstream.

Hm. I'm not sure I understand. What are `buildRustPackage` and `buildNpmPackage` doing then, if not import locks from fetched sources?

aMOPel commented:
But there is the `import from lock` functionality for both, where one doesn't need to specify any hash. It does exist. But you are saying it is bad practice?

Can you explain to me why it's bad or point me to some resources?

aMOPel commented:
Thank you  🙏 

Alright. So having the `import from lockfile` functionality is for Devs who want to use this build helper for their projects. The build with one hash is for nixpkgs.

Then I'm gonna provide both.

aMOPel commented:
@emilazy 

For further clarification:

I can analyse the lockfile in nix and change the properties of the derivation, just as long as it stays a single derivation (in this case a FOD). 

The problem arises when the number of derivations is dynamic depending on some nix evaluation.

aMOPel commented:
> If I understand your graph right, I don’t think you can have a derivation that produces one FOD output and one non‐FOD output. So annotating the JSON would have to be a separate derivation.

I guess my graph is a little misleading there. The single FOD fetcher just produces a directory of all the fechted files, as well as the annotated json file. So the hash of the FOD will also be over the annotated lockfile, but should not be a problem from what I understand.

> We probably don’t want the cache conversion to be a separate derivation from the actual build of the package, because it will bloat cache.nixos.org

This is a very interesting point :D good lord this is so much more complex, than I ever hoped for.

When you say

> “FOD that analyses the lock file, does all the fetching, and processes as little as possible” + “language‐specific builder that consumes the FOD output, converts it as appropriate, and does the build”

Does it mean this:

Instead of making the directory structure transformer an extra derivation,
it is just a build step when building the package?
The graph above is just about the fetcher. The result is fed into the package build.
So I can also just feed the results of either of the two fetcher variants into the package build itself, and have the directory structure transformer be a script, which is executed somewhen before the `buildPhase` of the package build.

> There may be Deno packages that have no lock file checked in upstream. Ideally we convince upstreams to fix this, but sometimes they’re unwilling. We would still like to be able to package these in Nixpkgs. Usually, what we do there is accept generating and vendoring a lock file in Nixpkgs for that case and do importCargoLock or equivalent. I think that your current design would not permit this, because the conversion derivation in the middle would be IFD.

Oh no :smile: So what you're saying is that there also needs to be a lockfile transformer equivalent in nix code.

With the architecture above, I wanted to reduce the duplication of the per language logic as much as possible.
After all, the single FOD fetcher and the "import from lock" fetcher both have basically the same logic, just in different languages and with slightly different outputs.

This duplication is 10 times worse for the deno fetcher, because there are so many edge cases and variations.
Writing and maintaining that code twice does not sound like fun.

So the way I see things now, I will only write the whole derivation build time pipeline. The nix eval time pipeline pretty much exists now, but there are some remaining issues, and the way I wrote it, it is not yet compatible with what is planned now. So I think I'm gonna cut it for now.

Does the adapted the graph look better to you?

![builder-architecture drawio](https://github.com/user-attachments/assets/1769fa8c-9b7e-41c4-b9d5-6890238ebc34)

aMOPel committed:
buildDenoPackage: init

Co-authored-by: rorosen

aMOPel committed:
buildDenoPackage: add tests

aMOPel committed:
deno: add output of denort binary

It's used by the denoBuildPackage build helper to compile deno projects
into a binary.

aMOPel committed:
doc: added docs for buildDenoPackage

aMOPel commented:
@hsjobeki
@philiptaron
@winterqt
@GetPsyched
@fricklerhandwerk
@ofalvai
@06kellyjac
@emilazy 

**This is a reopen of the reverted PR https://github.com/NixOS/nixpkgs/pull/407434**

This PR is ready for review. 

You don't need to read the PR comments above, as that was mainly about me understanding nixpkgs' requirements and emilazy explaining them to me

You can build the tests locally with 
```sh
nix build -f . --option allow-import-from-derivation false tests.build-deno-package
```
but mind that it will take a long time, since this PR adapts the `pkgs.deno` package, which means you have to fully rebuild it.

Compared to the previous PR, mainly the custom fetcher changed.

To understand what is going on in the fetcher, you can read this documentation I wrote about the features, the fetcher needs to have:

https://github.com/NixOS/nixpkgs/blob/bd54b5e47f0a7858336b38fa30aaea7541841eb5/pkgs/build-support/deno/fetch-deno-deps/readme.md

And to understand the general architecture of the fetcher, you can read this documentation, which is also a collection of general problems that language build helpers need to solve, along with a proposal for a standardized architecture, which could be improved upon and added to the nixpkgs documentation:

https://github.com/NixOS/nixpkgs/blob/bd54b5e47f0a7858336b38fa30aaea7541841eb5/pkgs/build-support/deno/fetch-deno-deps/generic-fetcher-architecture-docs.md

yacinehmito commented:
@aMOPel Were issues raised with Deno about the missing info in the lockfile?
(e.g. TypeScript declaration files provided in comments or in headers from esms.sh)

aorith commented:
First of all, thank you for your work @aMOPel 
I was trying to build [silverbullet](https://github.com/silverbulletmd/silverbullet/) on this commit https://github.com/silverbulletmd/silverbullet/commit/d77f0286d8eb43c241aba8597d29d7e09d51d7b3 using `buildDenoPackage` from this PR, but it seems that its not fetching all the dependencies:

```
 > error: Import 'https://deno.land/x/fuse@v6.4.1/dist/fuse.d.ts' failed
```

Running a `deno install` locally produces the following:

```sh
$ deno install --vendor --entrypoint ./silverbullet.ts
# . . .
$ ls -lrt vendor/deno.land/x/fuse@v6.4.1/dist/
total 24
-rw-r--r-- 1 aorith aorith 15426 ago  9 07:59 fuse.esm.min.js
-rw-r--r-- 1 aorith aorith  6894 ago  9 07:59 fuse.d.ts
```

Whereas if I add the same debug `ls` on `deno-build-hook.sh` the content is:

```
       > total 16
       > -rw-r--r-- 1 nixbld nixbld 15426 Aug  9 06:04 fuse.esm.min.js
```

aMOPel commented:
It is imported in a `@deno-types` declaration (now called `@ts-types`)
```
web/fuse_search.ts
1:// @deno-types="https://deno.land/x/fuse@v6.4.1/dist/fuse.d.ts"
```

This is very frustrating to add as a feature, because these types are only in the source files, not in the lock, but the fetcher already needs to know about them.

I could either pass the source files along with the lock file to the fetcher, but this creates problems with the updating of the hash.

Right now there is a check that tells you that the lock file has changed and you need to update the hash of the denoDeps, which will then result in a refetch.

Doing the same thing for the whole source code would mean refetching on every code change while developing.

Instead what I need is another derivation, before the fetching step, that takes the src code as input and returns a json with all the `@ts-types` urls in the source code (just by using grep). this json is passed to the fetcher and the urls in there are also fetched. the json can then be checked every time when building and if it changes, you need to update the deps hash.

As far as I understand this would not create an IFD so this should be fine.

aMOPel committed:
buildDenoPackage: add support for special type imports that can be
missing from the lockfile

aorith commented:
I get the same `denoDepsHash` as you.

I did a quick test and I'm not sure if this is required for a proper build without the type checking error, but after adding the following to `denoCompileFlags`:

```
    "--include=./build_plugs.ts"
    "--include=./build_bundle.ts"
    "--include=./build_web.ts"
    "--include=./silverbullet.ts"
```

It complains again about a dependency:

```
error: Import 'https://deno.land/x/esbuild@v0.25.1/mod.d.ts' failed.
```

Which indeed is not present:

```
-rw-r--r-- 1 nixbld nixbld 74643 Aug 11 17:58 mod.js
```

```
$ deno install --vendor --entrypoint ./silverbullet.ts ./build_web.ts ./build_bundle.ts ./build_plugs.ts
# . . .

$ rg mod.d
vendor/deno.land/x/esbuild@v0.25.1/mod.js
1:/// <reference types="./mod.d.ts" />
```

(The build requires `git` on `nativeBuildInputs`)

aorith commented:
@aMOPel No reason at all! I actually thought that it was something that you would like to add to the builder. I still cannot build the package, even with `--no-check`, but that's probably a mix of my limited knowledge and the project's somewhat unfriendly structure towards nix.

aMOPel commented:
> Since the PR descriptions says "this is in ready to merge state"

It says "ready for review", not merge.

In fact I plan to take out a little bit of code in the fetcher again as i mentioned in a previous comment, probably next week.

> could you give another simple overview of the final architecture

The two markdown files that I linked in the description try to do that. One explains the architecture and its reasoning. One explains the features that the fetcher has and doesn't have, the technical details of the deno dependency system.

I can try to rewrite them if they are not up to your standard.

> maybe we decide to split this into sub-prs so its easier for us to review?

I don't see how, but I'm open to be suprised.

You can split out the documentation, sure.

But it doesn't really make sense to split the code into pieces, since the build helper simply needs all the pieces to function. Also then you would need to write new tests for the pieces individually. Currently there are only e2e tests.

hsjobeki commented:
First of all, thank you for the huge amount of work you’ve put into this PR — it’s clear you’ve invested a lot of thought and effort. What I’m saying below comes from a place of wanting to get this merged in a timely way, rather than having it sit around for months because it’s too big to review properly.

That said, a PR of this size is very challenging for us to review effectively. Even with its good documentation, 5k lines of code in one go is almost unreviewable, and it increases the risk of bugs slipping in unnoticed. For us to maintain quality, we really need changes to be structured in smaller chunks.

You mentioned that the build helper needs all the pieces to function and has only e2e tests — but from our perspective, this suggests the code could benefit from more modularity. Smaller, well-defined units with corresponding unit tests would not only make it easier for us to review but also make the system more maintainable in the long term.

One possible path would be (as you already suggested):

- Extract the documentation as its own PR (as you suggested).
- Break the implementation into logical modules that can be tested and merged independently.
- Start with the core building block(s), with unit tests, and then layer additional functionality in follow-up PRs.

If we keep everything bundled as one huge change, the community won’t be able to give it the careful review it deserves.

hsjobeki commented:
Looking at the [architecture graph](
https://github.com/NixOS/nixpkgs/blob/bd54b5e47f0a7858336b38fa30aaea7541841eb5/pkgs/build-support/deno/fetch-deno-deps/generic-fetcher-architecture-docs.md) i have a few immediate questions:

- Does this PR introduce two different lockfile transformers and two different fetchers?
- Why is the architecture document not part of this PR? Including it here would make it much easier for us (and future maintainers) to understand and maintain the builder, rather than having the plan only in a separate commit on your fork.

- I find the build/eval-time line a bit confusing — though this may just be a matter of taste. Could you clarify how you see that boundary?

- What do you mean by “Nix realm”?
- What is `common lockfile` is that a custom format just for this pr? Sounds like the transformer(s) could have a spec. and unit tests.
- Some terms would benefit from clarity. I.e. `dependency cache` -> is this the deno cache?
- You could make all term specific to your problem, `language-package-manager` -> `deno`. finding a generic abstraction for all problems of the same kind is not necessarily a concern for this pr.

Also, since we’re actively discussing the design doc, it feels like it really should be added to this PR so the implementation and its rationale live together.

aMOPel commented:
> 5k lines

I believe 1.2k is lockfile. The one rust crate has a very big lockfile, and the other lockfiles are from the various e2e tests.

> Extract the documentation as its own PR (as you suggested).

Can do. I don't see how it makes the review easier, since the amount of code doesn't change, but can do.

> Break the implementation into logical modules that can be tested and merged independently. (seems the architecture was planned in a modular manner after all)

Yes. That would make the review easier. Unit tests would be good in a couple of places.

How many separate PRs were you thinking of?

Too much splitting would probably create a bunch additional effort. I would like to find a good balance.

I would like to split the builder and the fetcher and build tests for the multiple scripts, so the "lockfile transformer", "fetcher" and the two "file transformers". The tests would test each of those as a blackbox and just look at input and output. That way we also get a sort of spec for the formats.

Then I would create the fetcher PR first, then the builder PR and lastly the docs PR.

Does that sound acceptable?

> Does this PR introduce two different lockfile transformers and two different fetchers?

No, i tried to implement an "import from lockfile" functionality, but ultimately it was infeasible.

> Why is the architecture document not part of this PR?

It is. The links I posted lead to the markdown files from the very branch this pr stems from.

I just used permalinks so they wouldn't break. That doesn't mean the files are not on this branch anymore.

> I find the build/eval-time line a bit confusing — though this may just be a matter of taste. Could you clarify how you see that boundary?

> What do you mean by “Nix realm”?

I talk about this differentiation, because  the build helper can not do any IFD. So I try to visualize what steps happen during the evaluation time of nix code (the nix realm) vs what happens inside a nix build container.

> What is common lockfile is that a custom format just for this pr

I believe I did not do a good job making that clear in the markdown files.

Yes. It is custom format. It stemmed from my effort generalizing build helper architecture. One point of it is, that we need to decouple the format that the fetcher consumes from deno's lockfile format. Since the latter is subject to change and we only want to break the fetcher as few times as possible, to avoid refetching.


> Some terms would benefit from clarity. I.e. dependency cache -> is this the deno cache

Fair.

Yes. I believe so

> You could make all term specific to your problem, language-package-manager -> deno. finding a generic abstraction for all problems of the same kind is not necessarily a concern for this pr.

Yes. I believe what is missing in the documentation, is how this particular build helper's architecture relates to the general one. I can add that.

wanderer commented:
we are getting a build error now in https://github.com/masslbs/safewallet-to-beancount

```
      > Running phase: unpackPhase
       > Running phase: patchPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > no configure script, doing nothing
       > Running phase: buildPhase
       > error: Uncaught (in promise) Error: Invalid name@version format (version is required) in: "string" in string "string_decoder@1.3.0"
       >     throw new Error(`Invalid name@version format (version is required) in: "${split[0]}" in string "${fullString}"`);
       >           ^
```

is the underscore in string_decoder causing a problem?
